Avoid strdup in Stat::track

daschuer opened the pull request:
This avoids memory stacking in developer mode.

This fixes not a real memory leak, but it prevents out of memory conditions. 
By allocating a tag string for every 22 x 1 << 20 pipe entries. If we assume 30 byte for each tag, it will grow to 660 MB 

This is in addition to the preallocated pipes for StatReport which has a size of 32 byte consuming 704 MB

I do not think this will fix https://bugs.launchpad.net/mixxx/+bug/1448630, 
but it removed a noise from the memory view, to allow to track down the root issue.

daschuer committed:
Use static const  QStrings instead of string literals
this avoids allocating memory and unicode conversion over and over again

daschuer committed:
remove statistics from VSyncThread because it does no generate usefull info

daschuer committed:
deactivated Event stats in some threads to reduce the momory usage in developer mode

daschuer committed:
Removed statistics from EngineMaster
we have the values impicit by the outher scope statistics in SoundDevicePortAudio

daschuer commented:
I have added some additional commits to reduce the memory usage in developer mode and avoid some temporary memory allocations and unicode conversions. 
The required memory was reduced to 430 MB using Deere skin.

rryan commented:
Just checking, you don't intend to merge this right? Commenting out all the stat tracking isn't a good solution.

rryan commented:
A better solution would be to communicate the const char* template and its arguments in the Stat message itself.

daschuer commented:
I have actually intended to merge this. 
I have not commented out ALL the stats tracking. I have tried to remove those which are not that informative for remote debugging and kept the really interesting ones.  
For Example the benefit of the event counter in a thread is minor compared to the huge amount of 32 MB RAM it requires per thread. 
My target was to keep the RAM Footprint blow 500 MB. This should allow Mixxx to run on a 1 GB device with the --developer flag in use. 
I have removed the Events entirely from the sync thread because it requires memory allocation even if the --developer flag is not set which prevents the desired exactness. The collected counts are useless for remote debugging anyway.

> A better solution would be to communicate the const char* template and its arguments in the Stat message itself and then format them in the StatsManager.

I do not understand what you mean here? 

I have switched to the QString version to prevent unicode conversion and to benefit from the implicit sharing nature of QString. 

We may improve this further by replacing the plain C fifo by a C++ template fifo which supports multi produces. This way do not need a new queue for every thread.  
Unfortunately I did not find a drop in replacement for a multi producer lock free fifo on the web. 

I have a working solution using https://bitbucket.org/KjellKod/lock-free-wait-free-circularfifo/overview but since it is not multi producer and sizeof(QString) > sizeof(*QString) the result was worse.

daschuer commented:
> const char* template

Do you mean this: 
template<const char* TAG> void Event() 

This is not supported in C++ nor in C++11

rryan commented:
> Do you mean this:
> 
> template<const char* TAG> void Event() 
> This is not supported in C++ nor in C++11

Nope, was thinking something like a tagged union:

```c++
struct FormattedString {
  enum Type {
    NO_ARGUMENT,
    INT_ARGUMENT,
    CONST_CHAR_ARGUMENT
  }
  Type type;
  const char* tmpl;
  union {
    int int_arg;
    const char* str_arg;
  }
};
```

rryan commented:
> We may improve this further by replacing the plain C fifo by a C++ template fifo which supports multi produces. This way do not need a new queue for every thread.

I don't see the need to get rid of the per-thread fifo. Did you find that to be a performance issue?

daschuer commented:
> I don't see the need to get rid of the per-thread fifo. Did you find that to be a performance issue?

No in real, its a question of memory footprint. If we agree that a maximum memory footprint for Mixxx of 500 MB is reasonable to avoid swapped out memory on old systems, we just can't effort 32 MB for each of the 22 pipes. 
A common multi producer pipe with only 1 x 32 MB or a bit more will help here a lot.  
As a band aid I have just removed the most unimportant stats to reduce the number of required pipes.

daschuer commented:
> FormattedString

This will help as well but focuses the CPU performance. 
This PR mainly focuses the size of the memory footprint and the slowly gowning of it, which hides the other Windows memory issues we probably have. 

The Question is:
passing QSting* or char* 

char* is ideal for string literals, no need to strdup. But I have failed to distinguish string literals from other string pointers. There are some hacks described on the web, but non of them fit to our problem.  

So my conclusion was that a "static const QSting*" tag behaves not to bad because of the implicit sharing of the string itself. 
Hover my solution only reduces the footprint, it does not get rid of the the heap allocation for combined strings and it suffers some initial mallocs until every tag is cached.  

One additional thought: 
The combined strings are IMHO only uses as a tags. Maybe we can get rid of them by moving them to the static context.

daschuer commented:
@rryan: Do you have any updates or a conclusion?

daschuer commented:
@rryan: Any updates? This PR seams to become more important since we have a critical 1.12 bug report: 
https://bugs.launchpad.net/mixxx/+bug/1495317

daschuer commented:
This should be IMHO merged. Maybe someone else can have a critical eye on this?

rryan commented:
I agree with the goal of the PR but not the implementation (commenting out the code). I don't see why my proposal in https://github.com/mixxxdj/mixxx/pull/600#issuecomment-108925338 doesn't accomplish the goal of this PR?

daschuer commented:
Since https://github.com/mixxxdj/mixxx/pull/2355 is merged, the main issue of this PR is fixed. 
I have rebased the remaining performance improvements to master. 
@uklotzde: can you have a look?

uklotzde commented on src/util/event.h:
```diff
@@ -26,9 +26,34 @@ class Event {
+    static bool start(const char*) {
+        // this should not be uses to avoid unicode encoding and
+        // mamory alloc at every call. Use:
+        // static const QString tag(const char*);
```
A comment should explain that those function solely exist as a placeholder to prevent the invocation of QString(const char*) at compile time and must not be removed!

daschuer replied:
or use c++11 = delete

uklotzde commented on src/util/stat.h:
```diff
@@ -121,6 +121,17 @@ class Stat {
                       double value);
+
+  private:
+    static bool track(const char *,
```
Same here, i.e. intended compile-time error

uklotzde commented on src/vinylcontrol/vinylcontrolprocessor.cpp:
```diff
@@ -82,7 +82,7 @@ void VinylControlProcessor::run() {
 
     while (!m_bQuit) {
-        Event::start("VinylControlProcessor");
+        //Event::start("VinylControlProcessor");
```
This will not compile now after removing `//`. Please delete all those unused events.

daschuer committed:
Use c++11 delete keyword to disalow implicite string conversion

daschuer committed:
Remove commented code

daschuer commented:
Done.

uklotzde commented on src/util/event.h:
```diff
@@ -26,9 +26,18 @@ class Event {
     }
+
+    // Disallow to use this class with implicit converted char strings.
+    // This should not be uses to avoid unicode encoding and mmory
```
typo

uklotzde commented on src/util/stat.h:
```diff
@@ -121,6 +121,17 @@ class Stat {
+                      Stat::ComputeFlags,
+                      double) {
+        // this should not be uses to avoid unicode encoding and
+        // mamory alloc at every call. Use:
```
typo

uklotzde reviewed (commented):
Remaining typos, otherwise LGTM.

daschuer committed:
delete track(const char *, ...) and fix typos

daschuer commented:
AppVeyor and Visual studio has timed out.
Thank you for review.
